### PR TITLE
strictly regression of Hash#reject in Ruby 2.1.1

### PR DESCRIPTION
**Pull request by marutosi**

### CODE REVIEW

File: config/initializers/10-patches.rb [hunk 1]
```diff
@@ -233,6 +233,48 @@ class Errors
...
+# * https://www.ruby-lang.org/en/news/2014/03/10/regression-of-hash-reject-in-ruby-2-1-1/
+# * https://github.com/rails/rails/issues/14188
+# * https://github.com/rails/rails/pull/14198/files
+# * https://bugs.ruby-lang.org/issues/9576
+
+ruby_ver = RUBY_VERSION.split('.')
+if RAILS_VERSION < "4" &&
+    ruby_ver[0] >= 2 &&
+    (ruby_ver[1] == 1 && RUBY_PATCHLEVEL < "79" ||
+     ruby_ver[1] >= 2)
```
Comment (linki) [inline config/initializers/10-patches.rb:254]: why not just `RUBY_VERSION == "2.1.1"`?

Comment (marutosi) [inline config/initializers/10-patches.rb:254]: > why not just RUBY_VERSION == "2.1.1"?

https://bugs.ruby-lang.org/projects/ruby-21/repository/revisions/45227/diff/

Comment (linki) [inline config/initializers/10-patches.rb:254]: yeah, that explains it very well :+1:

**Comment by linki** [comment]
If we do this I would go with something like:

```
if Rails.version < "4"
  if RUBY_VERSION >= "2.2" || RUBY_VERSION == "2.1.1"
    apply_awesome_patch!
    warn "you might want to change how you do stuff" if RUBY_VERSION >= "2.2"
  end
else
  warn "this patch is not required anymore..."
end
```

more readable and still correct, i think.

**Comment by marutosi** [comment]
http://www.redmine.org/projects/redmine/repository/revisions/13115

**Comment by linki** [comment]
why did you close the PR, @marutosi?

**Comment by marutosi** [comment]
Your project does not support Ruby 1.8.

**Comment by linki** [comment]
what's [your patch](http://www.redmine.org/projects/redmine/repository/revisions/13115/diff/trunk/config/initializers/10-patches.rb) got to do with Ruby 1.8?

**Comment by marutosi** [comment]
http://www.redmine.org/builds/index.html
https://travis-ci.org/marutosi/redmine/builds/24176214

**Comment by linki** [comment]
you probably mean we don't support 2.0 and lower. but then there's still the `Rails::VERSION::MAJOR < 4` guard. why don't we need that?

**Comment by marutosi** [comment]
Just as you like.

**Comment by marutosi** [comment]
I have already started Rails 4 porting.
http://www.redmine.org/issues/14534

But it seems you have not.

**Comment by linki** [comment]
no, we haven't. how is your Rails 4 port going?

**Comment by marutosi** [comment]
http://marutosi.bitbucket.org/redmine-shinagawa-20120519/one-html/html/index.html#rails3-1

**Comment by linki** [comment]
link to some code please

**Comment by marutosi** [comment]
http://www.redmine.org/issues/14534

**Comment by linki** [comment]
that's not code.

**Comment by marutosi** [comment]
http://www.redmine.org/issues/14534#note-15

**Comment by linki** [comment]
Toshi MARUYAMA authored on Jan 21

**Comment by marutosi** [comment]
screenshot-marutosi - redmine-bb-rails4-20140526 - bitbucket - mozilla firefox

**Comment by marutosi** [comment]
screenshot-gitg - redmine finn-rails3

**Comment by linki** [comment]
why do you want everybody to interpret your answers?
give me a link to a repository along with a branch in plain text.

**Comment by marutosi** [comment]
Git branch is pointer to specific revision.
Do you understand git?

**Comment by marutosi** [comment]
$ git checkout -b are-you-kidding 064007e5cb0504db6f192
$ git log -n 1
commit 064007e5cb0504db6f1920fb59f3396172f65c16
Author: Toshi MARUYAMA
Date:   Tue Jan 21 22:18:41 2014 +0900

```
Rails4: Gemfile: use Rails4 github 4-0-stable branch

--HG--
branch : rails4.20140526-0
```

**Comment by marutosi** [comment]
screenshot-rails4 gemfile use rails4 github 4-0-stable branch 064007e marutosi-redmine - mozilla firefox

**Comment by linki** [comment]
give me a link to a repository along with the branch you are working on, so I can follow your progress.

**Comment by linki** [comment]
and not as a picture

**Comment by marutosi** [comment]
http://www.redmine.org/issues/14534#note-15

**Comment by linki** [comment]
:disappointed:

**Comment by marutosi** [comment]
<pre>
$ hg clone -U https://bitbucket.org/marutosi/redmine-bb-rails4-20140526
$ cd redmine-bb-rails4-20140526/
$ hg branches
rails4.20140526-0          10968:92689ce415c3
default                    10917:0f0beffca47b (inactive)
$ hg log -b rails4.20140526-0 --template='{rev} {node|short} {date|isodatesec}\n'
10968 92689ce415c3 2014-01-21 22:18:41 +0900
10967 d30aa90792e9 2014-05-24 21:58:57 +0900
10966 4c850ee840e5 2014-05-24 20:38:22 +0900
10965 178113827964 2014-05-24 15:11:04 +0900
10964 cedd7e778c28 2014-01-08 17:28:19 +0900
10963 5a5500f3dcf6 2014-01-19 01:27:22 +0900
10962 454213031aea 2013-12-22 16:19:43 +0900
10961 cec00c1575fe 2014-01-09 17:48:14 +0900
10960 403d6ab7af8e 2014-01-09 17:48:14 +0900
10959 fcb889197795 2014-01-07 22:07:32 +0900
10958 6a8e40dbbfb4 2014-01-26 15:57:30 +0900
10957 35bfc5025843 2014-01-08 23:12:22 +0900
10956 1c7451120a42 2014-01-10 11:25:03 +0900
10955 f488909999dd 2014-01-09 22:11:20 +0900
10954 a648ed4f6dd1 2014-01-10 01:00:20 +0900
10953 a12d6a321567 2014-01-09 23:50:30 +0900
10952 f86eba2c7382 2013-12-26 18:24:32 +0900
10951 5c425023101d 2013-12-23 13:12:49 +0900
10950 1d2a063d05ad 2014-01-08 12:36:50 +0900
10949 a945a565c199 2013-12-23 13:09:07 +0900
10948 2143133cf9bb 2013-12-30 23:09:11 +0900
10947 32d86547a6ad 2014-01-07 20:35:56 +0900
10946 0e52cab651cf 2014-01-07 18:21:49 +0900
10945 928cd2d68282 2013-12-26 10:21:18 +0900
10944 6659b75445ea 2013-12-22 16:53:52 +0900
10943 43a4b93a0008 2013-12-22 17:00:27 +0900
10942 b516f73e1f56 2013-12-20 22:58:50 +0900
10941 c7bdee292342 2013-12-22 18:51:23 +0900
10940 34e585045ac8 2013-12-26 11:49:50 +0900
10939 dbb1c4a4e5f3 2013-12-22 16:45:14 +0900
10938 431863c0b286 2013-12-22 16:41:53 +0900
10937 2feebedded0b 2013-12-22 15:52:12 +0900
10936 1fe689b05c08 2014-02-04 13:56:13 +0900
10935 312fa7592443 2013-12-20 22:58:50 +0900
10934 d2b303b35faf 2014-05-25 22:19:16 +0900
10933 e80eb3fac409 2014-05-25 20:40:22 +0900
10932 1a925a5e377f 2014-01-19 12:12:40 +0900
10931 38f5337cb334 2014-02-13 16:26:06 +0900
10930 b6eecc775f6c 2014-01-27 01:08:25 +0000
10929 85b3d65d8873 2014-01-27 01:08:09 +0000
10928 4a895cafe4d6 2014-01-27 01:07:52 +0000
10927 73623a56c88f 2014-01-27 01:07:34 +0000
10926 ecee1b94a0c0 2014-01-27 01:07:17 +0000
10925 fc614c4b185d 2014-01-27 01:06:58 +0000
10924 edcc52ecb24d 2014-01-27 14:00:45 +0900
10923 ebe83a911c3c 2014-01-27 13:42:16 +0900
10922 07a7d30cd1a1 2014-01-27 13:38:42 +0900
10921 651faf8fa76c 2014-01-27 13:22:59 +0900
10920 b9640c15b688 2013-12-21 01:39:32 +0900
10919 d72884c1efec 2014-05-25 22:12:42 +0900
10918 ba16afd80ede 2014-05-26 01:28:38 +0900
$ hg update rails4.20140526-0
1886 files updated, 0 files merged, 0 files removed, 0 files unresolved
</pre>

**Comment by marutosi** [comment]
1372c5cc-e7fc-11e3-8c26-6c1d27a381bc

**Comment by marutosi** [comment]
cb45edfa-e7fc-11e3-8266-b5743c2da9b0

**Comment by linki** [comment]
is there a reason you don't just post https://bitbucket.org/marutosi/redmine-bb-rails4-20140526/branch/rails4.20140526-0 and save us all the images?

**Comment by marutosi** [comment]
It is obvious.
http://www.redmine.org/issues/14534#note-15

We got many feedback on Rails3 portng.
http://www.redmine.org/issues/4796

If you cannot understand it, your DVCS skill is very poor.

**Comment by linki** [comment]
that answers everything :relieved: timestamped repository names with timestamped branches must be another skill i'm missing.

**Pull request closed without merging**